### ISSUE TITLE
Could defaults be adjusted based on the number of cores?

### ISSUE THREAD

**Issue by achamayou** [tlaplus/tlaplus#833]
There are a number of TLC [parameters](https://github.com/tlaplus/tlaplus/blob/master/general/docs/current-tools.md#java-system-properties), that are recommended for use on machines with multiple cores. These machines are very common these days, both in general and among TLC users.

Would it be acceptable to add code [detecting](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#availableProcessors()) that the machine has enough cores to justify using these optimisations, and to enable them by default?

I think this would improve user experience, and I'm happy to look into making the change if it is deemed acceptable.

**Comment by lemmy** [comment]
Historically, fine-tuning TLC was handled by the Toolbox. However, since the Toolbox is on the way out, we need to find a new home for this functionality. Regrettably, my previous attempt to integrate tuning into TLC a few years ago failed because certain VM parameters, like the GC settings, cannot be adjusted from within the VM itself.  @Calvin-L What are your thoughts?

**Comment by achamayou** [comment]
@lemmy the GC setting is definitely outside and difficult without a wrapper that at least sets `JAVA_TOOL_OPTIONS`, but `Dtlc2.tool.fp.FPSet.impl=tlc2.tool.fp.OffHeapDiskFPSet` and `-Dtlc2.tool.ModelChecker.BAQueue=true` are presumably relatively straightforward?

**Comment by lemmy** [comment]
Perhaps, we should scrutinize, i.e., peer-review BAQueue and simply make it the default.  As to OffHeapDiskFPSet, it suffers from long initialization times on Windows: https://github.com/tlaplus/tlaplus/issues/227

**Comment by achamayou** [comment]
@lemmy ah I did not know that. Perhaps the underlying platform can be used to make decisions about defaults too.

**Comment by lemmy** [comment]
https://github.com/tlaplus/tlaplus/issues/227#issuecomment-454478808 has a JVM patch that was never applied bc the JVM's sun.misc.Unsafe is dead.

**Comment by Calvin-L** [comment]
> Perhaps, it is finally time to bite the bullet and provide a wrapper script for TLC.

Yes, I think so.

**Comment by lemmy** [comment]
I'd assume that [tla-bin](https://github.com/pmer/tla-bin/blob/master/bin/tlc) is pretty much all we need. How do we prevent feature creep?  For example, downloading tla2tools.jar, installing a JVM, or [generating config files](https://github.com/hwayne/tlacli) should *not* be functionality of the wrapper.

**Comment by kape1395** [comment]
Maybe an `asdf` (https://asdf-vm.com/) plugin would be a good approach?

**Comment by jonesmartins** [comment]
UX-wise, maybe [tlacli](https://github.com/hwayne/tlacli) is a better alternative compared to tla-bin?

**Comment by achamayou** [comment]
@lemmy agree that _generating_ config files seems excessive, but being able to override/augment parts of the config with simple flags would be quite nice (tweak simulation runtime, add debug invariant, bump term limit etc).

**Comment by achamayou** [comment]
Yes, except perhaps the one-off enablement of debug invariants? I suppose it would be possible to do something like `${ADDITIONAL_INVARIANT:True}`.

**Comment by lemmy** [comment]
```
PROPERTIES
    ${DEBUG_INVARIANTS:InvA InvB InvC}
```

```shell
DEBUG_INVARIANTS="InvA InvB InvC InvD" java -jar tla2tools.jar ...
```

```shell
DEBUG_INVARIANTS="TRUE" java -jar tla2tools.jar ...
```
